### DISCUSSION TITLE
Dependencies in docs/requirements.txt are not found

### DISCUSSION THREAD

**Discussion by jachinte** [pyscaffold/pyscaffold#662] [Q&A] [answered]
Hi,

Where should I specify docs dependencies? I tried customizing the docs theme but the theme wasn't found.
These are the steps I followed:

1. Run `putup mypackage`
2. Edit `install_requires` in `setup.cfg` to include `furo`, the theme I want to use
3. Edit `docs/requirements.txt` to include `furo`
4. Edit `docs/conf.py` to set `html_theme = "furo"`
5. Run `tox -e docs`

This is what I get:

```
docs inst-nodeps: /.../mypackage/.tox/.tmp/package/1/mypackage-0.0.post1.dev2+g50f3cbc.tar.gz
docs installed: alabaster==0.7.12,attrs==22.1.0,Babel==2.10.3,certifi==2022.6.15,charset-normalizer==2.1.0,coverage==6.4.2,docutils==0.19,idna==3.3,imagesize==1.4.1,importlib-metadata==4.12.0,iniconfig==1.1.1,Jinja2==3.1.2,MarkupSafe==2.1.1,mypackage @ file:///.../mypackage/.tox/.tmp/package/1/mypackage-0.0.post1.dev2%2Bg50f3cbc.tar.gz,packaging==21.3,pluggy==1.0.0,py==1.11.0,Pygments==2.12.0,pyparsing==3.0.9,pytest==7.1.2,pytest-cov==3.0.0,pytz==2022.1,requests==2.28.1,snowballstemmer==2.2.0,Sphinx==5.1.1,sphinxcontrib-applehelp==1.0.2,sphinxcontrib-devhelp==1.0.2,sphinxcontrib-htmlhelp==2.0.0,sphinxcontrib-jsmath==1.0.1,sphinxcontrib-qthelp==1.0.3,sphinxcontrib-serializinghtml==1.1.5,tomli==2.0.1,urllib3==1.26.11,zipp==3.8.1
docs run-test-pre: PYTHONHASHSEED='943185761'
docs run-test: commands[0] | sphinx-build --color -b html -d /.../mypackage/docs/_build/doctrees /.../mypackage/docs /.../mypackage/docs/_build/html
Running Sphinx v5.1.1
...
Theme error:
no theme named 'furo' found (missing theme.conf?)
...
ERROR:   docs: commands failed
```

I tried running `putup --venv-install furo --update .` but I got the same error. Now, if I manually edit the `tox.ini` file, and add `furo` to the docs environment, it works:

```
[testenv:{docs,doctests,linkcheck}]
...
deps =
    -r {toxinidir}/docs/requirements.txt
    furo
...
```

```
tox -e docs
docs recreate: /.../mypackage/.tox/docs
docs installdeps: -r/.../mypackage/docs/requirements.txt, furo
...
docs: commands succeeded
congratulations :)
```

My question is why installing the package from `{toxinidir}/docs/requirements.txt` did not work? is this a bug?

Thank you!

--

I'm running these commands with Python 3.9.13 and pip 22.1.1 on a Mac.

**Comment by abravalheri** [comment, marked as answer]
Hi @jachinte, have you tried running `tox -re docs` when after you modified `docs/requirements.txt`?
(Currently `tox` has a limitation: it cannot automatically detect that a dependency file has changed and automatically identify that the cache has expired -- the `-r` flag forces the recreation of the environment).

---
I run the experiment below with Python 3.8, PyScaffold 4.3, tox 3.25.1 in a Ubuntu 20.04.4 LTS installation and I did not find the problem you are describing (maybe it is a Mac or a 3.9 specific problem?).
So I am assuming that maybe you had previously run `tox -e docs` which caused tox to cache the packages... Would this be the case?

```bash
cd /tmp
putup mypackage
cd mypackage
echo "furo" >> docs/requirements.txt
sed -i 's/alabaster/furo/' docs/conf.py
tox -e docs
```

If you think it is not an expired caching problem, could you share with us a sequence of commands (like the one below) that can be used to reproduce the problem?

**Comment by jachinte** [comment]
Hi @abravalheri,
Thank you for the quick response. Indeed, the problem was related to the limitation you mentioned.
After appending the `r` switch, the command to generate the docs ran successfully. 
I also tried the steps you ran on Ubuntu, and it works on Mac too.

Thanks again!
